core/src/main/java/org/infinispan/marshall/core/GlobalMarshaller.java

```diff
@@ -370,8 +414,189 @@ private void writeNonNullableObject(Object obj, BytesObjectOutput out) throws IO
       }
    }
 
+   private void writeArray(Class<?> clazz, Object array, BytesObjectOutput out) throws IOException {
+      out.writeByte(ID_ARRAY);
+      Class<?> componentType = clazz.getComponentType();
+
+      int length = Array.getLength(array);
+      boolean singleType = true;
+      Class<?> elementType = null;
+      int flags;
+      if (length == 0) {
+         flags = FLAG_ARRAY_EMPTY;
+      } else {
+         if (length <= Primitives.SMALL_ARRAY_MAX) {
+            flags = FLAG_ARRAY_SMALL;
+         } else if (length <= Primitives.MEDIUM_ARRAY_MAX) {
+            flags = FLAG_ARRAY_MEDIUM;
+         } else {
+            flags = FLAG_ARRAY_LARGE;
+         }
+         Object firstElement = Array.get(array, 0);
+         if (firstElement != null) {
+            elementType = firstElement.getClass();
+         }
+         for (int i = 1; i < length; ++i) {
```

Comment:
@galderz I was thinking about optimistically assuming that all instances are of the same type, and the most compact implementation (I can think of) would require one byte (1-0xFF) telling how many instances of single type there. The byte would be only reserved ahead, and then written once we:
a) reach end of array
b) reach different class instance
c) write 255 elements.
If we find out that 256th element is of the same type, we would set the single-type-length byte to 0 and reserve another byte to continue.
When hitting different class instance, I would back off to writing all the other elements without any compression, because this algorithm would add one byte per element for alternating sequences (sucha as when adding key on index `2*i` and value on `2*i + 1`) could be common.

The downside is one byte more per message, and increased complexity. Do you think it's worth it?